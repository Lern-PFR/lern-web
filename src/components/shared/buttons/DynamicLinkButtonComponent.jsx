import { cloneElement } from 'react';
import styled from 'styled-components';
import {
	border,
	color,
	display,
	flexbox,
	layout,
	margin,
	padding,
	position,
} from 'styled-system';
import PropTypes from 'prop-types';
import { disabled as disabledStyle } from 'theme/buttonStyles';

const StyledDynamicLinkButtonComponent = styled('button')(
	{
		cursor: 'pointer',
		padding: '0',
		textDecoration: (({ txtDecoration }) => txtDecoration || 'initial'),
		'&:hover': {
			color: ((props) => props.color.darker1),
			background: ((props) => props.bg.darker1),
		},

		'& *': {
			cursor: 'pointer',
		},
	},
	border,
	color,
	display,
	flexbox,
	layout,
	margin,
	padding,
	position,
);

/**
 * @name DynamicLinkButtonComponent
 * @description A dynamic component used to display styled link button element.
 *
 * @author TimothÃ©e Simon-Franza
 *
 * @param {Link}	children		: Link component to display inside the button.
 * @param {bool}	[disabled]		: Whather the button is disabled.
 * @param {object}	[linkStyle]		: Optional link style object.
 * @param {string}	targetUrl		: The url to redirect to.
 */
const DynamicLinkButtonComponent = ({ children, disabled, linkStyle, ...otherProps }) => {
	const { padding: paddingStyle, ...props } = otherProps;
	const { lineHeight, ...linkStyleProps } = linkStyle;

	return (
		<StyledDynamicLinkButtonComponent {...props} {...(disabled ? disabledStyle : {})}>
			{cloneElement(children, { disabled, padding: paddingStyle, ...linkStyleProps })}
		</StyledDynamicLinkButtonComponent>
	);
};

DynamicLinkButtonComponent.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
	disabled: PropTypes.bool,
	linkStyle: PropTypes.object,
};

DynamicLinkButtonComponent.defaultProps = {
	disabled: false,
	linkStyle: {},
};

export default DynamicLinkButtonComponent;
