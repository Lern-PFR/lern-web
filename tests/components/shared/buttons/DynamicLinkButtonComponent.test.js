import { shallow } from 'enzyme';
import DynamicLinkButtonComponent from 'components/shared/buttons/DynamicLinkButtonComponent';
import { cleanup } from '@testing-library/react';

describe('DynamicLinkButtonComponent', () => {
	afterEach(() => {
		cleanup();
	});

	it('should match previous snapshot', () => {
		const sut = (<DynamicLinkButtonComponent>Test button</DynamicLinkButtonComponent>);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});

	it('should match previous snapshot for disabled state', () => {
		const sut = (<DynamicLinkButtonComponent disabled>Test button</DynamicLinkButtonComponent>);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});
});
