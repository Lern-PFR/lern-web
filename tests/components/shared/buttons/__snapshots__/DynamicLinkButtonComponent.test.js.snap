// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicLinkButtonComponent should match previous snapshot 1`] = `
<styled.button>
  <Component
    disabled={false}
  />
</styled.button>
`;

exports[`DynamicLinkButtonComponent should match previous snapshot for disabled state 1`] = `
<styled.button
  bg="F6F6F4"
  color="#8C8C8C"
>
  <Component
    disabled={true}
  />
</styled.button>
`;
